cmd/expand-pool.go

```diff
@@ -0,0 +1,561 @@
+	PoolExpandStatusStartDecommission               PoolExpandStatus = "PoolExpandStatusStartDecommission"
+	PoolExpandStatusWaitForDecommissionComplete     PoolExpandStatus = "PoolExpandStatusWaitForDecommissionComplete"
+	PoolExpandStatusWaitForRemoveOldDataDir         PoolExpandStatus = "PoolExpandStatusWaitForRemoveOldDataDir"
+	PoolExpandStatusWaitForRenameNewDataDir         PoolExpandStatus = "PoolExpandStatusWaitForRenameNewDataDir"
+	PoolExpandStatusWaitForCorrectGlobalEndpoints   PoolExpandStatus = "PoolExpandStatusWaitForCorrectGlobalEndpoints"
+	PoolExpandStatusComplete                        PoolExpandStatus = "PoolExpandStatusComplete"
+)
+
+const (
+	poolExpandStatusFile = "expand.json"
+)
+
+type selfPoolExpand struct {
+	FileMD5 string `json:"fileMD5"`
+	// empty means the pool is not expanded
+	// PoolExpandStatusComplete means the pool is expanded and ready for another expand
+	Status    PoolExpandStatus `json:"status"`
+	PoolIndex int              `json:"poolIndex"`
+}
+
+func nextPoolExpandStatus(ps PoolExpandStatus) PoolExpandStatus {
+	switch ps {
+	case "":
+		return PoolExpandStatusWaitForFileChange
+	case PoolExpandStatusWaitForFileChange:
+		return PoolExpandStatusWaitForPreGlobalEndpointsChange
+	case PoolExpandStatusWaitForPreGlobalEndpointsChange:
+		return PoolExpandStatusStartDecommission
+	case PoolExpandStatusStartDecommission:
+		return PoolExpandStatusWaitForDecommissionComplete
+	case PoolExpandStatusWaitForDecommissionComplete:
+		return PoolExpandStatusWaitForRemoveOldDataDir
+	case PoolExpandStatusWaitForRemoveOldDataDir:
+		return PoolExpandStatusWaitForRenameNewDataDir
+	case PoolExpandStatusWaitForRenameNewDataDir:
+		return PoolExpandStatusWaitForCorrectGlobalEndpoints
+	case PoolExpandStatusWaitForCorrectGlobalEndpoints:
+		return PoolExpandStatusComplete
+	case PoolExpandStatusComplete:
+		return ""
+	}
+	return ""
+}
+
+func loadPoolExpandStats() (*selfPoolExpand, error) {
+	localPaths := globalEndpoints.LocalDisksPaths()
+	filemd5 := ""
+	poolIndex := -1
+	statsMap := map[PoolExpandStatus]int{}
+	for _, lp := range localPaths {
+		pe := selfPoolExpand{}
+		body, err := os.ReadFile(path.Join(lp, minioMetaBucket, poolExpandStatusFile))
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			}
+			return nil, err
+		}
+		err = json.Unmarshal(body, &pe)
+		if err != nil {
+			return nil, err
+		}
+		if filemd5 == "" {
+			filemd5 = pe.FileMD5
+		}
+		if pe.FileMD5 != filemd5 {
+			return nil, fmt.Errorf("inconsistent pool expand status")
+		}
+		if poolIndex == -1 {
+			poolIndex = pe.PoolIndex
+		}
+		if pe.PoolIndex != poolIndex {
+			return nil, fmt.Errorf("inconsistent pool expand status")
+		}
+		// if empty, set to PoolExpandStatusComplete
+		// but why it's empty
+		if pe.Status == "" {
+			pe.Status = PoolExpandStatusComplete
+		}
+		statsMap[pe.Status]++
+	}
+	stat := ""
+	maxCount := -1
+	for s, c := range statsMap {
```

Comment:
not a big fan of one letter var names. better use `for status, count := range statsMap`